homeassistant/components/panasonic_viera/media_player.py

```diff
@@ -149,98 +95,212 @@ def name(self):
+    async def async_media_next_track(self):
+        """Send the fast forward command."""
+        await self._remote.async_send_key(Keys.fast_forward)
 
-    def play_media(self, media_type, media_id, **kwargs):
+    async def async_media_previous_track(self):
+        """Send the rewind command."""
+        await self._remote.async_send_key(Keys.rewind)
+
+    async def async_play_media(self, media_type, media_id, **kwargs):
+        """Play media."""
+        await self._remote.async_play_media(media_type, media_id)
+
+
+class Remote:
+    """The Remote class. It stores the TV properties and the remote control connection itself."""
+
+    def __init__(
+        self, hass, host, port, app_id=None, encryption_key=None,
+    ):
+        """Initialize the Remote class."""
+        self._hass = hass
+
+        self._host = host
+        self._port = port
+        self._app_id = app_id
+        self._encryption_key = encryption_key
+
+        self.state = None
+        self.available = False
+        self.volume = 0
+        self.muted = False
+        self.playing = True
+
+        self._control = None
+
+    async def async_create_remote_control(self, during_setup=False):
+        """Create remote control."""
+        control_existed = self._control is not None
+        try:
+            params = {}
+            if self._app_id and self._encryption_key:
+                params["app_id"] = self._app_id
+                params["encryption_key"] = self._encryption_key
+
+            self._control = await self._hass.async_add_executor_job(
+                partial(RemoteControl, self._host, self._port, **params)
+            )
+
+            self.state = STATE_ON
+            self.available = True
+        except (TimeoutError, URLError, OSError) as err:
+            if control_existed or during_setup:
+                _LOGGER.error("Could not establish remote connection: %s", err)
+                self._control = None
+                self.state = STATE_OFF
+                self.available = False
+        except Exception as err:  # pylint: disable=broad-except
+            if control_existed or during_setup:
+                _LOGGER.error("An unknown error occurred: %s", err)
+                self._control = None
+                self.state = STATE_OFF
+                self.available = False
+
+    async def async_update(self):
+        """Retrieve the latest data."""
+        if self._control is None:
+            await self.async_create_remote_control()
+            return
+
+        try:
+            self.muted = self._control.get_mute()
+            self.volume = self._control.get_volume() / 100
+
+            self.state = STATE_ON
+            self.available = True
+        except EncryptionRequired:
+            _LOGGER.error("The connection couldn't be encrypted")
+        except (TimeoutError, URLError, OSError):
+            self.state = STATE_OFF
+            self.available = False
+            await self.async_create_remote_control()
```

Comment:
Please make a helper method that has the try... except handling since that is the same for five methods.
```suggestion
        await self._handle_errors(self._update)
    
    def _update(self):
        """Update media_player state."""
        self.muted = self._control.get_mute()
        self.volume = self._control.get_volume() / 100
        self.state = STATE_ON
        self.available = True
    
    async def _handle_errors(self, func, *args)
        """Handle errors from func, set available and reconnect if needed."""
        try:
            await self.hass.async_add_executor_job(func, *args)
        except EncryptionRequired:
            _LOGGER.error("The connection couldn't be encrypted")
        except (TimeoutError, URLError, OSError):
            self.state = STATE_OFF
            self.available = False
            await self.async_create_remote_control()
```

Reply:
Thank you so much for making the example (as you might have noticed, I'm not a Python expert, so I would have struggled with this for sure)